Sming/Arch/Esp32/Core/HardwarePWM.cpp

```diff
@@ -0,0 +1,303 @@
+		ledc_timer.clk_cfg = LEDC_AUTO_CLK;
+
+		debug_d("ledc_timer.\r\n"
+				"\tspeed_mode: %i\r\n"
+				"\ttimer_num: %i\r\n"
+				"\tduty_resolution: %i\r\n"
+				"\tfreq: %i\n\tclk_cfg: %i\r\n\n",
+				ledc_timer.speed_mode, ledc_timer.timer_num, ledc_timer.duty_resolution, ledc_timer.freq_hz,
+				ledc_timer.clk_cfg);
+		ESP_ERROR_CHECK(ledc_timer_config(&ledc_timer));
+
+		/*
+		 * Prepare and then apply the LEDC PWM channel configuration
+		 */
+		ledc_channel.speed_mode = pinToGroup(i);
+		ledc_channel.channel = pinToChannel(i);
+		ledc_channel.timer_sel = pinToTimer(i);
+		ledc_channel.intr_type = LEDC_INTR_DISABLE;
+		ledc_channel.gpio_num = pins[i];
+		ledc_channel.duty = 0;
+		ledc_channel.hpoint = 0;
+		debug_d("ledc_channel\n"
+				"\tspeed_mode: %i\r\n"
+				"\tchannel: %i\r\n"
+				"\ttimer_sel %i\r\n"
+				"\tintr_type: %i\r\n"
+				"\tgpio_num: %i\r\n"
+				"\tduty: %i\r\n"
+				"\thpoint: %i\r\n\n",
+				pinToGroup(i), pinToChannel(i), pinToTimer(i), ledc_channel.intr_type, pins[i], 0, 0);
+		ESP_ERROR_CHECK(ledc_channel_config(&ledc_channel));
+		ledc_bind_channel_timer(pinToGroup(i), pinToChannel(i), pinToTimer(i));
+	}
+	maxduty = maxDuty(DEFAULT_RESOLUTION);
+}
+
+HardwarePWM::~HardwarePWM()
+{
+	// Stop pwm for all pins and set idle level to 0
+	for(uint8_t i = 0; i < channel_count; i++) {
+		ledc_stop(pinToGroup(i), pinToChannel(i), 0);
+	}
+}
+
+uint8_t HardwarePWM::getChannel(uint8_t pin)
+{
+	for(uint8_t i = 0; i < channel_count; i++) {
+		if(channels[i] == pin) {
+			// debug_d("getChannel %d is %d", pin, i);
+			return i;
+		}
+	}
+	return -1;
+}
+
+uint32_t HardwarePWM::getDutyChan(uint8_t chan)
+{
+	// esp32 defines the frequency / period per timer
+	return (chan == PWM_BAD_CHANNEL) ? 0 : ledc_get_duty(pinToGroup(chan), pinToChannel(chan));
+}
+
+bool HardwarePWM::setDutyChan(uint8_t chan, uint32_t duty, bool update)
+{
+	if(chan == PWM_BAD_CHANNEL) {
+		return false;
+	}
+
+	if(duty <= maxduty) {
+		ESP_ERROR_CHECK(ledc_set_duty(pinToGroup(chan), pinToChannel(chan), duty));
+		if(update) {
+			ESP_ERROR_CHECK(ledc_update_duty(pinToGroup(chan), pinToChannel(chan)));
+			//update();
+		}
+		return true;
+	}
+
+	debug_d("Duty cycle value too high for current period.");
+	return false;
+}
+
+uint32_t HardwarePWM::getPeriod()
+{
+	// Sming does not know how to handle different frequencies for channels: this will require an extended interface.
+	// For now, just report the period for group 0 channel 0.
+	return frequencyToPeriod(ledc_get_freq(ledc_mode_t(0), ledc_timer_t(0)));
+}
+
+void HardwarePWM::setPeriod(uint32_t period)
+{
+	// Set the frequency globally, will add per timer functions later.
+	// Also, this can be done smarter.
+	for(uint8_t i = 0; i < channel_count; i++) {
+		ESP_ERROR_CHECK(ledc_set_freq(pinToGroup(i), pinToTimer(i), periodToFrequency(period)));
+	}
+	// ledc_update_duty();
+	update();
+}
+
+void HardwarePWM::update()
+{
+	// ledc_update_duty();
```

Comment:
Aside from optimisation, this method should allow multiple channels to be reconfigured, then updated simultaneously to keep them synchronised. Unfortunately the IDF HAL doesn't support this so requires further work.

My personal preference here would be to strip away all the HAL code and implement using low-level calls (such as `ledc_ll_ls_channel_update`). We'd also need to track which channels have changed and update them as a group. Something to consider as a future update, perhaps?

Reply:
I see where you're coming from. I was trying to create new static structures holding state to keep things easy, but being able to synchronize a change does require that.
But then again, when I thought about all the channel/timer relationships, this could get a much richer thing anyway and would need more state to be handled. Maybe it's worth writing an intermediary abstraction to handle all that. I'll think about it

Reply:
so, thinking a bit more about the interface, there's a few tings that might be of interest.

- depending o the SoC, there are four or eight timers that feed into six to 16 pwm channels. 
- the timers are configurable for duty cycle resolution and output frequency 
- high speed timers allow for glitch free change
- the same basic pwm system is also used for motor control, which might require different frequencies

My primary motivation to work on this is that I build led controllers that do thee to five channel (RGB to RGBWWCW) pwm. Longer term, my goal is to have a firmware that can use all available pwm channels of an SOC but break it out into multilple lights (maybe five times RGB, four times RGBW or three times RGBWWCW). As for the update function: that to me would mean that I want to be able to update specific groups of channels with one call and leave the rest alone. Also, I would want to have the ability to configure all channels that belong to one light to use one timer while respecting the fact, that some motor control might use an altogether different set of pwm parameters.

On the API level, this would require some notion of grouping channels / pwm pins, which would extend the current API and possibly break compatibility with existing code and/or other platforms, hence my thought of implementing an intermediary abstraction.

On the HardwarePWM level, I would then implement all channels to use the same timer (or two, if HS block is available), thus making sure they are all running at the same frequency and bit width.
I'm thinking of introducing a concept of channel groups that would configure channel assignment to timers and timer parameters. This way, the HardwarePWM interface would (mostly) still be concerned with channels.
There are a few functions in HardwarePWM that would, by this definition, not technically set channel parameters but group parameters, such as get/setPeriod/Frequency and update. I would overload them with a group parameter.
If no group is explicitly configured, all channels would end up in one default group (two if HS_ channels are being used) and those calls would change the values for that group.
I'll try to put that together in a new branch.